workflowai/core/utils/_pydantic_test.py

```diff
@@ -0,0 +1,123 @@
+
+
+class TestConstructModelRecursive:
+    def test_simple_model(self):
+        class SimpleModel(BaseModel):
+            name1: str
+            name2: str
+
+        constructed = construct_model_recursive(SimpleModel, {"name1": "John"})
+        assert isinstance(constructed, SimpleModel)
+        assert constructed.name1 == "John"
+        with pytest.raises(AttributeError):
+            _ = constructed.name2
+
+    def test_list_of_strings(self):
+        class ListOfStrings(BaseModel):
+            strings: list[str]
+
+        constructed = construct_model_recursive(ListOfStrings, {"strings": ["a", "b"]})
+        assert isinstance(constructed, ListOfStrings)
+        assert constructed.strings == ["a", "b"]
+
+    @pytest.mark.parametrize(
+        "payload",
+        [
+            {},
+            {"field1": "hello"},
+            {"nested": {"name": "hello", "field2": "world"}},
+            {"nested": {"name": "hello"}},
+        ],
+    )
+    def test_nested_model(self, payload: dict[str, Any]):
+        class NestedModel(BaseModel):
+            name: str
+            field2: str
+
+        class OuterModel(BaseModel):
+            field1: str
+            nested: NestedModel
+
+        constructed = construct_model_recursive(OuterModel, payload)
+        assert isinstance(constructed, OuterModel), "constructed is not an instance of OuterModel"
+        if "field1" in payload:
+            assert isinstance(constructed.field1, str), "field1 is not a string"
+        else:
+            with pytest.raises(AttributeError):
+                _ = constructed.field1
+
+        if "nested" in payload:
+            assert isinstance(constructed.nested, NestedModel), "nested is not an instance of NestedModel"
+            if "name" in payload["nested"]:
+                assert isinstance(constructed.nested.name, str)
+            else:
+                with pytest.raises(AttributeError):
+                    _ = constructed.nested.name
+
+            if "field2" in payload["nested"]:
+                assert isinstance(constructed.nested.field2, str)
+            else:
+                with pytest.raises(AttributeError):
+                    _ = constructed.nested.field2
+        else:
+            with pytest.raises(AttributeError):
+                _ = constructed.nested
+
+    def test_list_of_models(self):
+        class NestedModel(BaseModel):
+            name: str
+            field2: str
+
+        class ListOfModels(BaseModel):
+            models: list[NestedModel]
+
+        constructed = construct_model_recursive(
+            ListOfModels,
+            {"models": [{"name": "hello", "field2": "world"}, {"name": "hello"}]},
+        )
+        assert isinstance(constructed, ListOfModels)
+        assert isinstance(constructed.models, list)
+        assert isinstance(constructed.models[0], NestedModel)
+        assert isinstance(constructed.models[1], NestedModel)
```

Comment:
So models[1] is a NestedModel without the "field2" attribute?

That would mean that accessing NestedModel.field2 on the chunk processing will raise an AttributeError. And this error can be predicted by the linter... So that will crash at runtime.

I don't have the perfect solution for that. Maybe we should enforce all output class to only have optionals in the newest SDK version ? If not raise an error even before starting the stream ?

This way field2 will be none instead of absent; and you are forced by the linter to handle the case where "field2" is None in the stream processing...

Reply:
We mention the caveat of non optional fields in the Readme.
There is no way to do "partial validation" in Pydantic unfortunately.

Another way to handle the issue would be to compute "healthy defaults" for each field but it might get quite complex.. I'll make a quick attempt.

Reply:
Ok 👌, thank you! <br>Having warnings when streaming an output with non-optional would be useful too.